Dockerfile for essentia with tensorflow support

albincorreya opened the pull request:
This PR adds the following changes:
- Dockerfile for the following base images and configuration
`ubuntu18.04` along with python3 bindings
`ubuntu19.04` along with python3 bindings

- Updated readme and Makefile

alastair commented on Dockerfile.ubuntu19.04-python3-tensorflow:
```diff
@@ -0,0 +1,31 @@
+    && pip3 install --no-cache-dir tensorflow==1.15.0
+
+RUN apt-get update \
+    && apt-get install -y build-essential git \
```
this is missing the installation of all of the -dev packages, which means none of the audioloader algorithms are compiled, and therefore none of the examples either.

When I add these packages, building the examples quits with an error:
```
→ building from /essentia/essentia/src/python
[302/443] Linking build/src/examples/essentia_standard_beatsmarker
13:33:23 runner ['/usr/bin/g++', '-pthread', 'src/examples/standard_beatsmarker.cpp.1.o', '-o/essentia/essentia/build/src/examples/essentia_standard_beatsmarker', '-Wl,-Bstatic', '-Wl,-Bdynamic', '-Lsrc', '-L/usr/local/lib', '-lessentia', '-lfftw3f', '-lavformat', '-lavcodec', '-lavutil', '-lavresample', '-lsamplerate', '-ltag', '-lyaml', '-lpywrap_tensorflow_internal', '-ltensorflow_framework']
/usr/local/lib/libpywrap_tensorflow_internal.so: undefined reference to `PyList_Sort'
/usr/local/lib/libpywrap_tensorflow_internal.so: undefined reference to `PyType_IsSubtype'
/usr/local/lib/libpywrap_tensorflow_internal.so: undefined reference to `PySys_WriteStdout'
/usr/local/lib/libpywrap_tensorflow_internal.so: undefined reference to `PyObject_HasAttrString'
.... lots more
```
it appears that because we link to pywrap_tensorflow_internal we also need to link against libpython, but this isn't happening.

palonso replied:
I believe that the correct way to expose `libpython` should be by adding a `Requires: python` line when we [define the pkg-config file from the .so contained in the Tensorflow wheels](
https://github.com/MTG/essentia/blob/0a17738f7f1ac5881a6e80c3095eb7f0f8736252/src/3rdparty/tensorflow/setup_tensorflow.py#L87), but this would require pkg-config to be aware of the `libpython` package in order to `waf configure` sucessfuly, which is the case in my Ubuntu machine but not in the `manylinux` docker images that we use to build the wheels. 

Linking against the Tensorflow libraries contained in the wheels is a hack to make both Essentia and TensorFlow importable from the same Python session but to me, it doesn't make much sense to have the executable examples linked to a Tensorflow library full of Python symbols that are never used and maybe we should not support it.

Why not just dividing the examples and Python bindings build into 2 separate steps? 

```
src/3rdparty/tensorflow/setup_from_libtensorflow.sh
waf configure --build-static --with-examples --with-vamp --with-tensorflow
waf install

waf clean

src/3rdparty/tensorflow/setup_from_python.sh
waf configure --build-static --with-python --with-tensorflow
waf install
```

@dbogdanov what is your vision on this?

dbogdanov replied:
We'll have to download and store tensorflow twice in this case (two versions of the shared library, 800GB), this may be way too much. 

After further discussion with @pabloEntropia, it seems that the best for now would be to patch the tensorflow.pc file in the Docker script after running `src/3rdparty/tensorflow/setup_from_python.sh` adding `Requires: python`.

In the long term, we should also resolve the incompatibility of TensorFlow and Essentia extensions when the latter one is linked to a non-python version of libtensorflow shared library (in this case it's impossible to import both tensorflow and essentia together in Python). We should file an issue on TensorFlow related to that.

albincorreya commented:
@alastair I just added the Dockerfile for nightly builds of essentia with tensorflow-support for ubuntu18.04 and ubuntu19.04 base images. Also updated the make file and readme. For properly linking the essentia examples, I'd done the patch to tensorflow.pc file by adding `Requires: python3` as @dbogdanov suggested above.  I tested the python3 bindings and essentia example binaries inside the locally-built docker containers and it was working as expected.

alastair commented on README.md:
```diff
@@ -30,10 +28,16 @@ These images are based off the [v2.1_beta5 release]((https://github.com/MTG/esse
 
   *  `latest` ([Dockerfile.ubuntu18.04-nightly](https://github.com/MTG/essentia-docker/blob/master/nightly/Dockerfile.ubuntu18.04-nightly))
 
+Following tags are available for using essentia with [tensorflow-support](https://mtg.github.io/essentia-labs//news/2020/01/16/tensorflow-models-released/).
```
Add this as a separate header for Tensorflow support. Put the link to the essentia-labs page here.

alastair commented on Makefile:
```diff
@@ -13,12 +13,17 @@ latest:
 nightly:
 	docker build --pull -f nightly/Dockerfile.ubuntu18.04-nightly . -t mtgupf/essentia-nightly:latest
+	docker build --pull -f nightly/Dockerfile.ubuntu18.04-tensorflow-nightly . -t mtgupf/essentia-nightly:ubuntu18.04-tensorflow
+	docker build --pull -f nightly/Dockerfile.ubuntu19.04-tensorflow-nightly . -t mtgupf/essentia-nightly:ubuntu19.04-tensorflow
```
I'd make this a separate tensorflow target. I registered essentia-tensorflow on docker hub too, so we can tag it as that.

albincorreya replied:
You mean having `essentia-tensorflow` in a seperate docker hub repo? There is already one here https://hub.docker.com/r/mtgupf/essentia-tensorflow which I made earlier for the demos. Should we use this repo to push the latest tensorflow support images or in the nightly one?

alastair replied:
Yes, exactly. Let's use `essentia-tensorflow`

alastair commented on README.md:
```diff
@@ -30,10 +28,16 @@ These images are based off the [v2.1_beta5 release]((https://github.com/MTG/esse
+
+  *  `ubuntu18.04-tensorflow` ([Dockerfile.ubuntu18.04-tensorflow-nightly](https://github.com/MTG/essentia-docker/blob/master/nightly/Dockerfile.ubuntu18.04-tensorflow-nightly))
+
+  *  `ubuntu19.04-tensorflow` ([Dockerfile.ubuntu19.04-tensorflow-nightly](https://github.com/MTG/essentia-docker/blob/master/nightly/Dockerfile.ubuntu19.04-tensorflow-nightly))
```
When we discussed this, we decided that we'd only release tensorflow support for LTS, in order to not have too many different packages. Do we need 19.04 too?

albincorreya replied:
Yes, that makes sense. I will remove the 19.04 one

albincorreya commented:
Hi @alastair, made the changes as you suggested here and also pushed the image to the docker hub with latest readme.

alastair commented:
can you put the dockerfile in a new directory? (like nightly and python2)

albincorreya commented:
> can you put the dockerfile in a new directory? (like nightly and python2)

@alastair Yes, it's already put it inside a dir called `tensorflow`. Is that what you meant?
https://github.com/MTG/essentia-docker/tree/essentia-tf

alastair commented:
oh, maybe I didn't refresh before I saw the diff. Thanks.

alastair commented:
This is great, thanks. Ready to merge. One small thing, I saw that you edited the `nightly/Dockerfile.ubuntu18.04-nightly` and the only thing you did was delete the newline at the end of the file...
Can you please rebase and force-push this branch to undo that change?

albincorreya commented:
> This is great, thanks. Ready to merge. One small thing, I saw that you edited the `nightly/Dockerfile.ubuntu18.04-nightly` and the only thing you did was delete the newline at the end of the file...

oh sorry just noticed, it happened while I was trying to fix the merge conflict in a hacky way.

> Can you please rebase and force-push this branch to undo that change?

Yes sure
